17_display_activities_from_backend

catbus00 opened the pull request:
### This PR resolves #17 

## Change Summary

- Create ActivitiesDataFetch.tsx file in Activities
- Update /activities to pull data from backend repo

## I Had Difficulty With

_Finding a way to work around with the CORS error from backend test server and learning to work with Axios._

catbus00 committed:
Create ActivitiesDataFetch.tsx file to fulfill the ticket  Display activities from backend#17

dargaCode committed:
Merge branch 'dev' into 17_display_activities_from_backend

catbus00 committed:
Update Routes

dargaCode committed:
Merge branch 'dev' into 17_display_activities_from_backend

dargaCode commented on package-lock.json:
Hrm what do you think is causing this file to change? Since you didn't add or update any dependency packages in `package.json` this shouldn't be changing at all. Can you please revert this file? You may be able to just delete it an then `npm install` again

catbus00 replied:
Every time I run install, it was updating to the latest dependencies. I've since reverted to the version before the commit so hopefully it's the same as everyone else on the team now.

dargaCode commented on src/components/ActivitiesList/ActivitiesDataFetch.tsx:
```diff
@@ -0,0 +1,52 @@
+const ActivitiesList: React.FC = (): ReactElement => {
+  const [activities, setActivities] = useState<Activity[]>([]);
+
+  useEffect(() => {
```
Nice! I know you said you were having trouble using Axios, and I should have mentioned that was just the example code that CtD provided us with - you can definitely use any library or just `fetch` if you prefer!

catbus00 replied:
I'll keep using Axios as I got it to work and it's great learning for me, if that's ok :)

dargaCode commented on src/components/ActivitiesList/ActivitiesDataFetch.tsx:
```diff
@@ -0,0 +1,52 @@
+            <p>Description: {activity.description}</p>
+            <p>Category: {activity.category}</p>
+            <p>Vote Count: {activity.votes}</p>
+          </div>
```
do you see Typescript errors in your IDE saying that `_id` and `votes` aren't valid keys for the `Activity` type? 

![Screen Shot 2023-12-03 at 5 51 44 PM](https://github.com/Code-the-Dream-School/ee-prac-team2-front/assets/12771651/24ffc046-3a03-468f-ba92-fc5a2b2b1af7)

When we want to add or modify data keys like this we have to go to the original `Activity` interface and update the `interface`!

in this case it looks like the names of what the API is delivering are slightly different than the shape we have here: 

![Screen Shot 2023-12-03 at 5 54 39 PM](https://github.com/Code-the-Dream-School/ee-prac-team2-front/assets/12771651/9f582864-0119-4bc3-85cb-cc42570eb2f9)

Possibly this was changed after you submitted your PR, but the data I'm seeing in the fetch has `_id` and an optional `votes` which is an object:

<img width="659" alt="Screen Shot 2023-12-03 at 5 59 07 PM" src="https://github.com/Code-the-Dream-School/ee-prac-team2-front/assets/12771651/9e5151e4-8c62-48a4-a18f-1a9fc512bd0c">

You can check with Mo but if this is the correct shape you'll have to update where your component is rendering the `voteCount` data from to point to this as well.

dargaCode commented on src/components/ActivitiesList/ActivitiesDataFetch.tsx:
```diff
@@ -0,0 +1,52 @@
+    <div className={styles.activitiesContainer}>
+      <h1>Activities</h1>
+      <div className={styles.activitiesList}>
+        {activities.map((activity) => (
+          <div key={activity._id} className={styles.activityItem}>
+            <p>Activity Name: {activity.name}</p>
+            <p>Description: {activity.description}</p>
```
Since you don't need to ever do anything with the `activity` object as a whole, try destyructuring its data in the parameter definition, so you use less of the dot notation:

```suggestion
        {activities.map({_id, name, description, votes} => (
          <div key={activity._id} className={styles.activityItem}>
            <p>Activity Name: {activity.name}</p>
            <p>Description: {activity.description}</p>
```

if you needed to pass the entire `activity` to a subcomponent etc, you could leave it as `activity` in the parameter, but then do the destructuring on the first line of the function body. 

Sometimes when things are complicated and a lot of data is flying around it's more readable to see a lot of dot notation, but in this case it's pretty straightforward so you can save some work

dargaCode commented on src/components/ActivitiesList/ActivitiesDataFetch.tsx:
```diff
@@ -0,0 +1,52 @@
+            <p>Activity Name: {activity.name}</p>
+            <p>Description: {activity.description}</p>
+            <p>Category: {activity.category}</p>
+            <p>Vote Count: {activity.votes}</p>
```
This also could have changed since you submitted your PR, but I'm seeing that sometimes `votes` is entirely `undefined`. Because of that we'll get an error if we try to access `undefined.count`, so we can use the [optional chaining operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) and n[ullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) to make sure the activity with no `votes` object displays a zero instead of crashing the page:

```suggestion
              <p>Vote Count: {activity.votes?.count ?? 0}</p>
```

`votes?.count` is a shortcut for what used to have to be done in a ternary before this feature: 

`votes !== undefined ? votes.count : undefined`

then once we have either a number or the `undefined`, the `??` (a more modern version of `||` fallback) allows the value to fall back to `0`

catbus00 replied:
The shape of the vote has changed @dargaCode based on the changes requested for the backend schema. Should I hold the PR until the backend team finalizes it first?

msrezaie replied:
This is an example of a resource received from the Activity data model:
```
{
          "_id": "656b62e0fa36aea592c2308e",
          "name": "Movie Night",
          "description": "Gather friends and watch a movie.",
          "category": "social",
          "createdAt": "2023-12-02T17:01:20.844Z",
          "updatedAt": "2023-12-02T17:01:20.844Z",
          "vote": {
              "_id": "656b62e0fa36aea592c2308f",
              "activityID": "656b62e0fa36aea592c2308e",
              "voters": [
                  {
                      "_id": "656b62e6fa36aea592c23093",
                      "name": "Mo",
                      "email": "mo@email.com",
                      "createdAt": "2023-12-02T17:01:26.417Z",
                      "updatedAt": "2023-12-02T17:01:26.417Z",
                      "__v": 0
                  },
                  {
                      "_id": "6568fb622d91faf5b3b777e1",
                      "name": "Morez",
                      "email": "example@email.com",
                      "createdAt": "2023-11-30T21:15:14.363Z",
                      "updatedAt": "2023-11-30T21:15:14.363Z",
                      "__v": 0
                  }
              ],
              "createdAt": "2023-12-02T17:01:20.844Z",
              "updatedAt": "2023-12-02T17:02:02.943Z",
              "__v": 2
          },
          "__v": 0
},
```

dargaCode replied:
Thanks for resolving this you two!

dargaCode commented on src/components/ActivitiesList/ActivitiesDataFetch.tsx:
Good job on this! 

There are some best practices we can follow for how our components are broken out and what they contain, which I'm sorry I didn't think to talk about before we started you on this task!

You've got a component with a different name than the file that contains it, and it's sortof "pretending" to be the component that we had before, while actually it's an entirely new component. 

I don't know if you remember [the `container`/`view` pattern ](https://reactpatterns.com/#container-component)that we looked at in the todo list. 

By creating a `container` component, and doing all our fetching there, we can leave the `view` or "presentational component" exactly how it was working when we were passing it mock data. It just renders whatever data it receives in the props, so it doesn't need to care if the data was hardcoded or fetched. 


- `<App />` renders new component` <ActivitiesContainer />` in the `<Route/>`. 
- `<ActivitiesContainer /> does all the fetching ( so it includes the `useState`, and the `useEffect` that you've written. 
- Then the `<ActivitiesContainer /> passes the fetched `activities` state into `<ActivitiesList />` as its prop. 

The original component shouldn't need any modifying at all except to update the shape of the `Activity` type as I mentioned in another comment. 







Instead of making an entirely new alternate component that does a fetch and returning it as `ActivitiesList`

dargaCode replied:
ps please move the `container` out to a new folder we can use for this and other future needs: 

`src/containers/ActivitiesContainer.tsx`

catbus00 replied:
Hi Darga, I've reorganized this! This was very helpful, thank you for the feedback :)

dargaCode replied:
Thanks Jillian! Do you have un-pushed changes on your local branch?

dargaCode reviewed (changes requested):
Great job Jillian! I'm sorry this took so long to review - I hadn't gotten my backend environment variables and stuff set up properly yet. 

The functionality is looking good here - just needs a couple small type updates. The structure of where you actually put this logic in the files is what I'm mainly requesting changes on. You should be able to rename `ActivitiesDataFetch` to `ActivitiesContainer`, and then leave the rendering where it was before in `ActivitiesList`

catbus00 committed:
Revert package.lock to original version as requested by @dargaCode

catbus00 committed:
Update containers location

catbus00 committed:
remove unused nodemon

catbus00 committed:
update server proxy port

catbus00 committed:
Update Activities interface based on http://localhost:8000/api/v1/activities

catbus00 committed:
update voters length

catbus00 commented:
@dargaCode yes, I just pushed the changes. Thank you!

catbus00 commented:
Hello, this PR has been corrected and ready for review. Thank you! :)

dargaCode commented on tsconfig.json:
```diff
@@ -21,6 +21,7 @@
     "paths": {
       "@assets/*": ["src/assets/*"],
       "@components/*": ["src/components/*"],
+      "@containers/*": ["src/containers/*"],
```
Awesome thank you!

dargaCode commented on vite.config.ts:
```diff
@@ -10,6 +10,7 @@ export default defineConfig({
     alias: {
       "@assets": path.resolve(__dirname, "./src/assets"),
       "@components": path.resolve(__dirname, "./src/components"),
+      "@containers": path.resolve(__dirname, "./src/containers"),
```
🙏

dargaCode commented on vite.config.ts:
```diff
@@ -18,7 +18,7 @@ export default defineConfig({
     proxy: {
       "/api": {
-        target: "http://localhost:8080",
+        target: "http://localhost:8000",
```
Oh damn thank you for noticing this! Was this causing errors before?

dargaCode committed:
Update package-lock.json

dargaCode committed:
Complete the Container/View pattern

I know I need to let you move on to the next task, so I'm going to merge in these changes.

I made the `ActivitiesContainer` component name match the filename, and then replaced the new rendering you added with just rendering the preexisting `ActivitiesList`

So this way there's no API logic in the "view" `ActivitiesList`, and there are no UI details inside of `ActivitiesContainer`

dargaCode reviewed (approved):
Hi Jillian, great job on this feature!

I'm going to check in a last couple of small changes here and then merge this, but please [take a look at what I changed](https://github.com/Code-the-Dream-School/ee-prac-team2-front/pull/37/commits/da71c24feb8e6448559269257b03306894c47245) and let me know your questions ❓❓❓
